microprofile/grpc/client/src/main/java/io/helidon/microprofile/grpc/client/GrpcChannelsProvider.java

```diff
@@ -0,0 +1,216 @@
+    public Channel channel(String name) {
+        if (name == null) {
+            throw new NullPointerException("name cannot be null.");
+        }
+        if (name.trim().isEmpty()) {
+            throw new IllegalArgumentException("name cannot be empty or blank.");
+        }
+        GrpcChannelDescriptor chCfg = channelConfigs.computeIfAbsent(name, hostName ->
+                GrpcChannelDescriptor.builder().host(name).build());
+        return createChannel(name, chCfg);
+    }
+
+    /**
+     * Creates a channel from a channel descriptor.
+     *
+     * @param name the channel name
+     * @param descriptor the channel descriptor
+     * @return the channel
+     */
+    private Channel createChannel(String name, GrpcChannelDescriptor descriptor) {
+        Tls clientTls = descriptor.tls().orElse(null);
+        if (clientTls == null) {
+            throw new IllegalArgumentException("Client TLS must be configured for gRPC proxy client");
+        }
+        int port = descriptor.port();
+        if (port <= 0) {
+            port = discoverServerPort(name);
+        }
+        GrpcClient grpcClient = GrpcClient.builder()
+                .tls(clientTls)
+                .baseUri("https://" + descriptor.host() + ":" + port)
+                .build();
+        return grpcClient.channel();
+    }
+
+    /**
+     * Used for unit testing: if port not set, try to obtain port from server CDI
+     * extension via reflection, without introducing a static dependency.
+     *
+     * @param name the channel name
+     * @return server port
+     * @throws java.lang.IllegalArgumentException if unable to discover port
+     */
+    @SuppressWarnings("unchecked")
+    private static int discoverServerPort(String name) {
+        try {
+            Class<? extends Extension> extClass = (Class<? extends Extension>) Class
+                    .forName("io.helidon.microprofile.server.ServerCdiExtension");
+            Extension extension = CDI.current().getBeanManager().getExtension(extClass);
+            Method m = extension.getClass().getMethod("port", String.class);
+            return (int) m.invoke(extension, new Object[] {"@default"});
+        } catch (ReflectiveOperationException e) {
+            throw new IllegalArgumentException("Unable to determine server port for channel " + name);
+        }
+    }
+
+    /**
+     * Builder builds an instance of {@link GrpcChannelsProvider}.
+     */
+    public static class Builder implements io.helidon.common.Builder<Builder, GrpcChannelsProvider> {
+
+        private final Map<String, GrpcChannelDescriptor> channelConfigs = new HashMap<>();
+
+        private Builder(Config config) {
+            // Add the default channel (which can be overridden in the config)
+            channel(DEFAULT_CHANNEL_NAME, GrpcChannelDescriptor.builder().build());
+
+            if (config == null) {
+                return;
+            }
+
+            Config channelsConfig = config.get(CFG_KEY_CHANNELS);
```

Comment:
This is an unusual usage of config, that is very hard to document right. 
Is there a way to do this with the standard usage of Helidon builders, or at least with appropriate config metadata documentation?

Reply:
I should be possible to convert this to a Helidon builder, I need to look into that.

Reply:
Pushed commit to update the configuration of channels.